--
-- PostgreSQL database dump
--

-- Dumped from database version 10.13
-- Dumped by pg_dump version 13.1

-- Started on 2021-05-22 16:04:11 CDT

SET statement_timeout = 0;
SET lock_timeout = 0;
SET idle_in_transaction_session_timeout = 0;
SET client_encoding = 'UTF8';
SET standard_conforming_strings = on;
SELECT pg_catalog.set_config('search_path', '', false);
SET check_function_bodies = false;
SET xmloption = content;
SET client_min_messages = warning;
SET row_security = off;

--
-- TOC entry 3994 (class 0 OID 16595)
-- Dependencies: 244
-- Data for Name: species; Type: TABLE DATA; Schema: geneweaver; Owner: odeadmin
--

INSERT INTO geneweaver.species (sp_id, sp_name, sp_taxid, sp_ref_gdb_id, sp_date, sp_biomart_info, sp_source_data) VALUES (0, '', 0, NULL, '2011-03-14', NULL, NULL);
INSERT INTO geneweaver.species (sp_id, sp_name, sp_taxid, sp_ref_gdb_id, sp_date, sp_biomart_info, sp_source_data) VALUES (1, 'Mus musculus', 10090, 10, '2011-03-14', 'mmusculus_gene_ensembl	mgi_id', '{"MGI_Coordinate.rpt": {"header": 1, "nulls": ["", "null"], "primary_source": "MGI_Coordinate.rpt", "columns": {"1": {"excludelist": ["Cytogenetic Marker", "DNA Segment", "Complex/Cluster/Region", "QTL", "BAC/YAC end", "Other Genome Feature"], "required": "yes", "mapto": "gi_type"}, "0": {"required": "yes", "mapto": "gi_accession"}, "3": {"mapto": "gi_name"}, "2": {"mapto": "gi_symbol"}, "5": {"mapto": "gi_chromosome"}, "7": {"mapto": "gi_end_bp"}, "6": {"mapto": "gi_start_bp"}, "8": {"mapto": "gi_strand"}}}, "dbxrefs": {"mgi": "MGI:"}, "urls": {"MGI_Coordinate.rpt": "ftp://ftp.informatics.jax.org/pub/reports/MGI_Coordinate.rpt"}}');
INSERT INTO geneweaver.species (sp_id, sp_name, sp_taxid, sp_ref_gdb_id, sp_date, sp_biomart_info, sp_source_data) VALUES (3, 'Rattus norvegicus', 10116, 12, '2011-03-14', 'rnorvegicus_gene_ensembl	rgd', '{"GENES_RAT.txt": {"header": 64, "prefix": "RGD", "nulls": [""], "primary_source": "GENES_RAT.txt", "columns": {"15": {"mapto": "gi_end_bp"}, "14": {"mapto": "gi_start_bp"}, "16": {"mapto": "gi_strand"}, "33": {"excludelist": "WITHDRAWN", "required": "yes", "mapto": "exclusion"}, "36": {"mapto": "gi_type"}, "1": {"mapto": "gi_symbol"}, "0": {"required": "yes", "mapto": "gi_accession"}, "3": {"mapto": "gi_description"}, "2": {"mapto": "gi_name"}, "6": {"mapto": "gi_chromosome"}}}, "dbxrefs": {"rgd": "RGD"}, "urls": {"GENES_RAT.txt": "ftp://rgd.mcw.edu/pub/data_release/GENES_RAT.txt"}}');
INSERT INTO geneweaver.species (sp_id, sp_name, sp_taxid, sp_ref_gdb_id, sp_date, sp_biomart_info, sp_source_data) VALUES (6, 'Macaca mulatta', 9544, 1, '2011-03-14', 'mmulatta_gene_ensembl	entrezgene', '{"seq_gene.q.gz": {"header": 1, "nulls": [""], "primary_source": "seq_gene.q.gz", "columns": {"1": {"mapto": "gi_type"}, "0": {"regex": "^GeneID:(?P<gi_accession>[0-9]+)$", "mapto": "gi_accession"}, "2": {"mapto": "gi_symbol"}, "7": {"mapto": "gi_name"}}}, "dbxrefs": {}, "urls": {"seq_gene.q.gz": "ftp://anonymous@ftp.ncbi.nih.gov/genomes/MapView/Macaca_mulatta/objects/current/initial_release/seq_gene.q.gz"}}');
INSERT INTO geneweaver.species (sp_id, sp_name, sp_taxid, sp_ref_gdb_id, sp_date, sp_biomart_info, sp_source_data) VALUES (8, 'Caenorhabditis elegans', 6239, 15, '2011-03-14', 'celegans_gene_ensembl	wormbase_gene', '{"biomart_8_1.txt": {"header": 0, "nulls": [""], "primary_source": "biomart_8_1.txt", "columns": {"1": {"mapto": "gi_symbol"}, "0": {"required": "yes", "mapto": "gi_accession"}, "3": {"mapto": "gi_description"}, "2": {"mapto": "gi_name"}, "5": {"mapto": "gi_start_bp"}, "4": {"mapto": "gi_chromosome"}, "7": {"mapto": "gi_strand"}, "6": {"mapto": "gi_end_bp"}}}, "dbxrefs": {"wormbase": ""}, "urls": {"biomart_8_1.txt": "http://caprica.caltech.edu:9002/biomart/martservice?query=%3C%3Fxml+version%3D%221.0%22+encoding%3D%22UTF-8%22%3F%3E%3C%21DOCTYPE+Query%3E%3CQuery++virtualSchemaName+%3D+%22WS220-testing%22+formatter+%3D+%22TSV%22+header+%3D+%220%22+uniqueRows+%3D+%220%22+count+%3D+%22%22+datasetConfigVersion+%3D+%220.6%22+%3E%3CDataset+name+%3D+%22wormbase_gene%22+interface+%3D+%22default%22+%3E%3CFilter+name+%3D+%22species_selection%22+value+%3D+%22Caenorhabditis+elegans%22%2F%3E%3CFilter+name+%3D+%22identity_status%22+value+%3D+%22Live%22%2F%3E%3CAttribute+name+%3D+%22gene%22+%2F%3E%3CAttribute+name+%3D+%22public_name%22+%2F%3E%3CAttribute+name+%3D+%22sequence_name%22+%2F%3E%3CAttribute+name+%3D+%22concise_description%22+%2F%3E%3CAttribute+name+%3D+%22chromosome_name%22+%2F%3E%3CAttribute+name+%3D+%22chromosome_start%22+%2F%3E%3CAttribute+name+%3D+%22chromosome_end%22+%2F%3E%3CAttribute+name+%3D+%22chromosome_strand%22+%2F%3E%3C%2FDataset%3E%3C%2FQuery%3E"}}');
INSERT INTO geneweaver.species (sp_id, sp_name, sp_taxid, sp_ref_gdb_id, sp_date, sp_biomart_info, sp_source_data) VALUES (2, 'Homo sapiens', 9606, 11, '2011-03-14', 'hsapiens_gene_ensembl	hgnc_id', '{"dbxrefs": {"hgnc": "HGNC:"}, "biomart_2_1.txt": {"header": 0, "prefix": "", "nulls": [""], "primary_source": "biomart_2_1.txt", "columns": {"1": {"mapto": "gi_symbol"}, "0": {"required": "yes", "mapto": "gi_accession"}, "3": {"regex": "(?P<gi_chromosome>^[^_]*$)", "required": "yes", "mapto": "gi_chromosome"}, "2": {"mapto": "gi_name"}, "5": {"mapto": "gi_end_bp"}, "4": {"mapto": "gi_start_bp"}, "7": {"excludelist": ["snRNA", "lincRNA", "Mt_rRNA", "rRNA", "tRNA_pseudogene", "miRNA", "LRG_gene", "snoRNA", "misc_RNA", "scRNA_pseudogene"], "required": "yes", "mapto": "gi_type"}, "6": {"mapto": "gi_strand"}}}, "urls": {"biomart_2_1.txt": "http://www.biomart.org/biomart/martservice?query=%3C%3Fxml+version%3D%221.0%22+encoding%3D%22UTF-8%22%3F%3E%3C%21DOCTYPE+Query%3E%3CQuery++virtualSchemaName+%3D+%22default%22+formatter+%3D+%22TSV%22+header+%3D+%220%22+uniqueRows+%3D+%220%22+count+%3D+%22%22+datasetConfigVersion+%3D+%220.6%22+%3E%3CDataset+name+%3D+%22hsapiens_gene_ensembl%22+interface+%3D+%22default%22+%3E%3CFilter+name+%3D+%22with_hgnc%22+excluded+%3D+%220%22%2F%3E%3CAttribute+name+%3D+%22hgnc_id%22+%2F%3E%3CAttribute+name+%3D+%22hgnc_symbol%22+%2F%3E%3CAttribute+name+%3D+%22hgnc_transcript_name%22+%2F%3E%3CAttribute+name+%3D+%22chromosome_name%22+%2F%3E%3CAttribute+name+%3D+%22start_position%22+%2F%3E%3CAttribute+name+%3D+%22end_position%22+%2F%3E%3CAttribute+name+%3D+%22strand%22+%2F%3E%3CAttribute+name+%3D+%22gene_biotype%22+%2F%3E%3C%2FDataset%3E%3C%2FQuery%3E"}}');
INSERT INTO geneweaver.species (sp_id, sp_name, sp_taxid, sp_ref_gdb_id, sp_date, sp_biomart_info, sp_source_data) VALUES (4, 'Danio rerio', 7955, 13, '2011-03-14', 'drerio_gene_ensembl	zfin_id', '{"genetic_markers.txt": {"header": 0, "nulls": [""], "primary_source": "genetic_markers.txt", "columns": {"1": {"mapto": "gi_symbol"}, "0": {"required": "yes", "mapto": "gi_accession"}, "3": {"excludelist": ["PAC_END", "SSLP", "ATB", "TGCONSTRCT", "PAC", "BAC", "RAPD", "ETCONSTRCT", "PTCONSTRCT", "GTCONSTRCT", "MRPHLNO", "STS", "FOSMID", "EFG", "REGION", "BAC_END"], "required": "yes", "mapto": "gi_type"}, "2": {"mapto": "gi_name"}}}, "zfin_gene.gff3": {"header": 1, "nulls": [""], "columns": {"0": {"mapto": "gi_chromosome"}, "3": {"mapto": "gi_start_bp"}, "2": {"mapto": "gi_type"}, "4": {"mapto": "gi_end_bp"}, "6": {"mapto": "gi_strand"}, "8": {"regex": "ID=(?P<gi_accession>[^;]+);", "required": "yes", "mapto": "gi_accession"}}}, "primary_source": "genetic_markers.txt", "dbxrefs": {"zfin": ""}, "urls": {"zfin_gene.gff3": "http://zfin.org/data_transfer/Downloads/zfin_gene.gff3", "genetic_markers.txt": "http://zfin.org/data_transfer/Downloads/genetic_markers.txt"}}');
INSERT INTO geneweaver.species (sp_id, sp_name, sp_taxid, sp_ref_gdb_id, sp_date, sp_biomart_info, sp_source_data) VALUES (5, 'Drosophila melanogaster', 7227, 14, '2011-03-14', 'dmelanogaster_gene_ensembl	flybase_gene_id', '{"gene_map_table_fb_2011_02.tsv.gz": {"header": 6, "nulls": [""], "primary_source": "gene_map_table_fb_2011_02.tsv.gz", "columns": {"1": {"mapto": "gi_accession"}, "0": {"regex": "^(?P<gi_symbol>[^\\\\]*)$", "required": "yes", "mapto": "gi_symbol"}, "4": {"regex": "^(?P<gi_chromosome>[0-9XYLR]*):(?P<gi_start_bp>[0-9]*)\\.\\.(?P<gi_end_bp>[0-9]*)\\((?P<gi_strand>-?1)\\)$", "mapto": "bymatch"}}}, "dbxrefs": {"flybase": ""}, "primary_source": "gene_map_table_fb_2011_02.tsv.gz", "gene_summaries.tsv.gz": {"header": 1, "nulls": [""], "columns": {"1": {"required": "yes", "mapto": "gi_description"}, "0": {"required": "yes", "mapto": "gi_accession"}}}, "urls": {"gene_map_table_fb_2011_02.tsv.gz": "http://flybase.org/static_pages/downloads/FB2013_04/genes/gene_map_table_fb_2013_04.tsv.gz", "gene_summaries.tsv.gz": "http://flybase.org/static_pages/downloads/FB2013_04/genes/gene_summaries.tsv.gz"}}');
INSERT INTO geneweaver.species (sp_id, sp_name, sp_taxid, sp_ref_gdb_id, sp_date, sp_biomart_info, sp_source_data) VALUES (11, 'Canis familiaris', 9615, 1, '2015-06-05', 'cfamiliaris_gene_ensembl	entrezgene', '{"canis_gene_info.txt": {"header": 1, "nulls": [""], "columns": {"1": {"mapto": "gi_accession"}, "8": {"mapto": "gi_description"}, "2": {"mapto": "gi_symbol"}, "9": {"mapto": "gi_type"}, "6": {"mapto": "gi_chromosome"}}}, "primary_source": "canis_gene_info.txt", "urls": {"canis_gene_info.txt": "ftp://anonymous@ftp.ncbi.nih.gov/genomes/MapView/Canis_lupus_familiaris/objects/current/initial_release/seq_gene.q.gz"}}');
INSERT INTO geneweaver.species (sp_id, sp_name, sp_taxid, sp_ref_gdb_id, sp_date, sp_biomart_info, sp_source_data) VALUES (10, 'Gallus gallus', 9031, 20, '2013-07-30', 'ggallus_gene_ensembl	entrezgene', '{"AGNC_Gallus_gallus.txt": {"header": 1, "prefix": "CGNC", "nulls": [""], "columns": {"0": {"mapto": "gi_accession"}, "3": {"mapto": "gi_symbol"}, "4": {"mapto": "gi_description"}}}, "primary_source": "AGNC_Gallus_gallus.txt", "dbxrefs": {"cgnc": "CGNC"}, "urls": {"AGNC_Gallus_gallus.txt": "http://www.agnc.msstate.edu/Downloads.aspx"}}');
INSERT INTO geneweaver.species (sp_id, sp_name, sp_taxid, sp_ref_gdb_id, sp_date, sp_biomart_info, sp_source_data) VALUES (9, 'Saccharomyces cerevisiae', 4932, 16, '2011-03-14', NULL, '{"SGD_features.tab": {"header": 0, "nulls": [""], "primary_source": "SGD_features.tab", "columns": {"11": {"mapto": "gi_strand"}, "10": {"mapto": "gi_end_bp"}, "15": {"mapto": "gi_description"}, "1": {"excludelist": ["CDEIII", "noncoding_exon", "binding_site", "rRNA", "insertion", "X_region", "telomeric_repeat", "ncRNA", "transposable_element_gene", "snRNA", "plus_1_translational_frameshift", "Y_element", "intron", "external_transcribed_spacer_region", "gene_cassette", "five_prime_UTR_intron", "Y_region", "Z2_region", "centromere", "CDEII", "ARS consensus sequence", "multigene locus", "mating_locus", "X_element_combinatorial_repeats", "telomere", "long_terminal_repeat", "W_region", "not physically mapped", "retrotransposon", "Z1_region", "ARS", "repeat_region", "tRNA", "CDEI", "non_transcribed_region", "internal_transcribed_spacer_region", "X_element_core_sequence", "snoRNA", "not in systematic sequence of S288C"], "required": "yes", "mapto": "gi_type"}, "0": {"required": "yes", "mapto": "gi_accession"}, "4": {"required": "yes", "mapto": "gi_symbol"}, "9": {"mapto": "gi_start_bp"}, "8": {"mapto": "gi_chromosome"}}}, "dbxrefs": {"sgd": "SGD"}, "urls": {"SGD_features.tab": "http://downloads.yeastgenome.org/chromosomal_feature/SGD_features.tab"}}');


--
-- TOC entry 4001 (class 0 OID 0)
-- Dependencies: 245
-- Name: species_sp_id_seq; Type: SEQUENCE SET; Schema: geneweaver; Owner: odeadmin
--

SELECT pg_catalog.setval('geneweaver.species_sp_id_seq', 11, true);


-- Completed on 2021-05-22 16:04:14 CDT

--
-- PostgreSQL database dump complete
--

